devlib/collector/perfetto.py

```diff
@@ -0,0 +1,68 @@
+#    Copyright 2023 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import time
+import subprocess
+
+from devlib.collector import (CollectorBase, CollectorOutput,
+                              CollectorOutputEntry)
+from devlib.exception import TargetStableError, HostError
+from devlib.utils.misc import check_output, which, memoized
+
+OUTPUT_PERFETTO_TRACE = 'trace.perfetto-trace'
+
+class PerfettoCollector(CollectorBase):
+
+    def __init__(self, target, config=None):
+        super(PerfettoCollector, self).__init__(target)
+        self.target_output_file = target.path.join(self.target.working_directory, OUTPUT_PERFETTO_TRACE)
+        self.output_path = None
+        self.config = config
+
+        if not self.target.is_installed('perfetto'):
+            raise TargetStableError('No perfetto found on device.')
+        self.target_binary = 'perfetto'
+
+    def start(self):
+        # start tracing and save the PID to stop it later
+        self.pid = self.target.execute(
```

Comment:
It may be a good idea to use the target.background() and the BackgroundCommand API to handle that instead of re-growing an ad-hoc version of it:

```
def start(self):
    ...
    if self.bg_cmd is None:
        self.bg_cmd = self.target.background(cmd)
    else:
        raise TargetStableError('Perfetto collector is not re-entrant'')
    ...

def stop(self):
    ...
    self.bg_cmd.kill()
    self.bg_cmd = None
    ...
```